src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -185,58 +188,37 @@ private static IReadOnlyList<TagHelperDescriptor> ResolveTagHelperDescriptors(Ge
-                // by the app to avoid some per-compilation costs.
-                // We determine if any of the reference assemblies have a newer timestamp than the output cache for the tag helpers. If not, we can re-use previously
-                // calculated results.
-                refTagHelpers = TagHelperSerializer.Deserialize(razorContext.RefsTagHelperOutputCachePath);
-            }
-            else
-            {
-                tagHelperFeature.DiscoveryFilter = TagHelperDiscoveryFilter.ReferenceAssemblies;
-                refTagHelpers = tagHelperFeature.GetDescriptors();
-                TagHelperSerializer.Serialize(razorContext.RefsTagHelperOutputCachePath, refTagHelpers);
-            }
-
-            tagHelperFeature.DiscoveryFilter = TagHelperDiscoveryFilter.CurrentCompilation;
+            var currentMetadataReference = GeneratorExecutionContext.Compilation.GetMetadataReference(GeneratorExecutionContext.Compilation.Assembly);
+            tagHelperFeature.TargetReference = currentMetadataReference;
             var assemblyTagHelpers = tagHelperFeature.GetDescriptors();
 
+            var refTagHelpers = GetTagHelperDescriptorsFromReferences(GeneratorExecutionContext.Compilation, tagHelperFeature);
+
             var result = new List<TagHelperDescriptor>(refTagHelpers.Count + assemblyTagHelpers.Count);
             result.AddRange(assemblyTagHelpers);
             result.AddRange(refTagHelpers);
 
             return result;
         }
 
-        private static DateTime? GetLastUpdatedReference(IEnumerable<MetadataReference> references)
+        private static IReadOnlyList<TagHelperDescriptor> GetTagHelperDescriptorsFromReferences(Compilation compilation, StaticCompilationTagHelperFeature tagHelperFeature)
         {
-            DateTime lastWriteTimeUtc = DateTime.MinValue;
+            List<TagHelperDescriptor> tagHelperDescriptors = new();
 
-            foreach (var reference in references)
+            foreach (var reference in compilation.References)
             {
-                // We expect all references in the compilation context to be backed by a file on disk. If not,
-                // we'll bail out and regenerate the tag helper cache where this is invoked.
-                if (reference is not PortableExecutableReference portableExecutableReference || string.IsNullOrEmpty(portableExecutableReference.FilePath))
+                var guid = reference.GetModuleVersionId(compilation) ?? Guid.NewGuid();
+                if (!_tagHelperCache.TryGetValue(guid, out var descriptors))
                 {
-                    return null;
+                    tagHelperFeature.TargetReference = reference;
+                    descriptors = tagHelperFeature.GetDescriptors();
+                    _tagHelperCache.AddOrUpdate(guid, descriptors, (key, currentValue) => descriptors);
```

Comment:
I think we should add some reasonable upper-bound to this. If your references keep changing this list grows unbounded. Also I would just do `_tagHelperCache[guid] = descriptors;`